forc-pkg/src/pkg.rs

```diff
@@ -1369,12 +1412,174 @@ pub fn fetch_git(fetch_id: u64, name: &str, pinned: &SourceGitPinned) -> Result<
+    }
+    Ok(found_local_repo)
+}
+
+/// Search and find the match repo between the given tag and locally available options
+fn find_repo_with_tag(
+    tag: &str,
+    package_name: &str,
+    checkouts_dir: PathBuf,
+) -> Result<Option<(PathBuf, String)>> {
+    let mut found_local_repo = None;
+    with_search_checkouts(
+        checkouts_dir,
+        package_name,
+        &mut |repo_index, repo_dir_path| {
+            // Get current head of the repo
+            let current_head = repo_index.head_with_time.0;
+            if let GitReference::Tag(curr_repo_tag) = repo_index.git_reference {
+                if curr_repo_tag == tag {
+                    found_local_repo = Some((repo_dir_path, current_head))
+                }
+            }
+            Ok(())
+        },
+    )?;
+    Ok(found_local_repo)
+}
+
+/// Search and find the match repo between the given rev and locally available options
+fn find_repo_with_rev(
+    rev: &str,
+    package_name: &str,
+    checkouts_dir: PathBuf,
+) -> Result<Option<(PathBuf, String)>> {
+    let mut found_local_repo = None;
+    with_search_checkouts(
+        checkouts_dir,
+        package_name,
+        &mut |repo_index, repo_dir_path| {
+            // Get current head of the repo
+            let current_head = repo_index.head_with_time.0;
+            if let GitReference::Rev(curr_repo_rev) = repo_index.git_reference {
+                if curr_repo_rev == rev {
+                    found_local_repo = Some((repo_dir_path, current_head))
+                }
+            }
+            Ok(())
+        },
+    )?;
+    Ok(found_local_repo)
+}
+
+/// Search local checkouts directory and apply the given function. This is used for iterating over
+/// possible options of a given package.
+fn with_search_checkouts<F>(checkouts_dir: PathBuf, package_name: &str, f: &mut F) -> Result<()>
+where
+    F: FnMut(GitSourceIndex, PathBuf) -> Result<()>,
+{
+    for entry in fs::read_dir(checkouts_dir)? {
+        let entry = entry?;
+        let folder_name = entry
+            .file_name()
+            .into_string()
+            .map_err(|_| anyhow!("invalid folder name"))?;
+        if folder_name.starts_with(package_name) {
+            // Search if the dir we are looking starts with the name of our package
+            for repo_dir in fs::read_dir(entry.path())? {
+                // Iterate over all dirs inside the `name-***` directory and try to open repo from
+                // each dirs inside this one
+                let repo_dir = repo_dir
+                    .map_err(|e| anyhow!("Cannot find local repo at checkouts dir {}", e))?;
+                if repo_dir.file_type()?.is_dir() {
+                    // Get the path of the current repo
+                    let repo_dir_path = repo_dir.path();
+                    // Get the index file from the found path
+                    let index_file = fs::read_to_string(repo_dir_path.join(".forc_index"))?;
```

Comment:
Rather than returning on `Err` here, lets just skip this repo if it doesn't have an index file. This should allow us to land this without users having to delete their old `.forc` directories, and should avoid returning an `Err` in the case that a repo has been checked out but the `.forc_index` file has not yet been written.